org.openhab.binding.zigbee/src/main/java/org/openhab/binding/zigbee/handler/ZigBeeSerialPort.java

```diff
@@ -279,15 +278,45 @@ public void serialEvent(SerialPortEvent event) {
         if (event.getEventType() == SerialPortEvent.DATA_AVAILABLE) {
             try {
                 synchronized (bufferSynchronisationObject) {
-                    int recv;
-                    while ((recv = inputStream.read()) != -1) {
-                        buffer[end++] = recv;
+                    int available = inputStream.available();
+                    logger.trace("Processing DATA_AVAILABLE event: have {} bytes available", available);
+                    byte[] buf = new byte[available];
+                    int offset = 0;
+                    while (offset != available) {
+                        if (logger.isTraceEnabled()) {
+                            logger.trace("Processing DATA_AVAILABLE event: try read  {} at offset {}",
+                                    available - offset, offset);
+                        }
+                        int n = inputStream.read(buf, offset, available - offset);
+                        if (logger.isTraceEnabled()) {
+                            logger.trace("Processing DATA_AVAILABLE event: did read {} of {} at offset {}", n,
+                                    available - offset, offset);
+                        }
+                        if (n <= 0) {
+                            logger.error("Processing DATA_AVAILABLE event: Expected to be able to read " + available
+                                    + " bytes, but saw error after " + offset);
+                            throw new IOException(
+                                    "Expected to be able to read " + available + " bytes, but saw error after "
+                                            + offset);
+                        }
+                        offset += n;
+                    }
+                    for (int i = 0; i < available; i++) {
+                        buffer[end++] = buf[i] & 0xff;
                         if (end >= RX_BUFFER_LEN) {
                             end = 0;
                         }
+                        if (end == start) {
+                            logger.warn("Processing DATA_AVAILABLE event: Serial buffer overrun");
```

Comment:
I'd suggest to drop this down to debug since the system will recover from this.

Reply:
> I'd suggest to drop this down to debug since the system will recover from this.

That's why warn is appropriate instead of error -  an unusual event, which may cause user visible anomalous behavior, and which may indicate more serious problems, but for which automatic recovery can be performed.

Reply:
Fine - let's leave this one, but please resolve the other one.

Reply:
I would be uncomfortable lowering the priority.

Reply:
As I said, let's leave this as it is.